x-pack/filebeat/input/awss3/collector.go

```diff
@@ -348,67 +353,79 @@ func (c *s3Collector) createEventsFromS3Info(svc s3iface.ClientAPI, info s3Info,
 		return err
 	}
 
 	if isS3ObjGzipped {
-		gzipReader, err := gzip.NewReader(reader)
+		gzipReader, err := gzip.NewReader(bodyReader)
 		if err != nil {
 			c.logger.Error(fmt.Errorf("gzip.NewReader failed for '%s' from S3 bucket '%s': %w", info.key, info.name, err))
 			return err
 		}
-		reader = bufio.NewReader(gzipReader)
-		gzipReader.Close()
+		defer gzipReader.Close()
+		bodyReader = bufio.NewReader(gzipReader)
 	}
 
 	// Decode JSON documents when content-type is "application/json" or expand_event_list_from_field is given in config
 	if resp.ContentType != nil && *resp.ContentType == "application/json" || info.expandEventListFromField != "" {
-		decoder := json.NewDecoder(reader)
+		decoder := json.NewDecoder(bodyReader)
 		err := c.decodeJSON(decoder, objectHash, info, s3Ctx)
 		if err != nil {
-			c.logger.Error(fmt.Errorf("decodeJSONWithKey failed for '%s' from S3 bucket '%s': %w", info.key, info.name, err))
-			return err
+			return fmt.Errorf("decodeJSONWithKey failed for '%s' from S3 bucket '%s': %w", info.key, info.name, err)
 		}
 		return nil
 	}
 
 	// handle s3 objects that are not json content-type
+	encodingFactory, ok := encoding.FindEncoding(c.config.Encoding)
+	if !ok || encodingFactory == nil {
+		return fmt.Errorf("unable to find '%v' encoding", c.config.Encoding)
+	}
+	enc, err := encodingFactory(bodyReader)
+	if err != nil {
+		return fmt.Errorf("failed to initialize encoding: %v", err)
+	}
+	var r reader.Reader
+	r, err = readfile.NewEncodeReader(ioutil.NopCloser(bodyReader), readfile.Config{
+		Codec:      enc,
+		BufferSize: c.config.BufferSize,
+		Terminator: c.config.LineTerminator,
+		MaxBytes:   c.config.MaxBytes * 4,
+	})
+	r = readfile.NewStripNewline(r, c.config.LineTerminator)
+
+	if c.config.Multiline != nil {
+		r, err = multiline.New(r, "\n", c.config.MaxBytes, c.config.Multiline)
+		if err != nil {
+			return fmt.Errorf("error setting up multiline: %v", err)
+		}
+	}
+
+	r = readfile.NewLimitReader(r, c.config.MaxBytes)
+
 	var offset int64
 	for {
-		log, err := readStringAndTrimDelimiter(reader)
+		message, err := r.Next()
 		if err == io.EOF {
-			// create event for last line
-			offset += int64(len(log))
-			event := createEvent(log, offset, info, objectHash, s3Ctx)
-			err = c.forwardEvent(event)
-			if err != nil {
-				c.logger.Error(fmt.Errorf("forwardEvent failed: %w", err))
-				return err
-			}
-			return nil
-		} else if err != nil {
-			c.logger.Error(fmt.Errorf("readStringAndTrimDelimiter failed: %w", err))
-			return err
+			// No more lines
+			break
 		}
-
-		if log == "" {
-			continue
+		if err != nil {
+			return fmt.Errorf("Error reading message: %v", err)
 		}
-
-		// create event per log line
-		offset += int64(len(log))
-		event := createEvent(log, offset, info, objectHash, s3Ctx)
+		event := createEvent(string(message.Content), offset, info, objectHash, s3Ctx)
+		offset += int64(message.Bytes)
 		err = c.forwardEvent(event)
 		if err != nil {
```

Comment:
Can you combine this into a one-liner, e.g. `if err = c.forwardEvent(event); err != nil {`.